Add documentation on multicompartment models

sarakonradi opened the pull request:
This PR explains the interpretation of the delay in NEST with regard to multicompartment neurons. It also updates ``pp_cond_exp_mc_urbanczik.h`` and ``urbanczik_synapse.h`` to RST format.

Many thanks to @jstapmanns for adding the description and figure.

pnbabu commented on models/pp_cond_exp_mc_urbanczik.h:
```diff
@@ -107,82 +107,137 @@ class pp_cond_exp_mc_urbanczik_parameters
+consideration. In NEST, the delay is at least one simulation time step and is
+assumed to be located entirely at the postsynaptic side. For point neurons, it
+represents the time it takes for an incoming spike to travel along the
+postsynaptic dendrite before it reaches the soma, see panel a). Conversely, if
```
It would be good to reference the figure wherever `panel` is mentioned.

sarakonradi replied:
Good idea!

pnbabu commented on models/urbanczik_synapse.h:
```diff
@@ -40,51 +40,66 @@
+Wmin        real   Minimum allowed weight
+=========   ====   =========================================================
+
+All other parameters are stored in in the neuron models that are compatible
```
```suggestion
All other parameters are stored in the neuron models that are compatible
```

pnbabu commented on models/pp_cond_exp_mc_urbanczik.h:
```diff
@@ -107,82 +107,137 @@ class pp_cond_exp_mc_urbanczik_parameters
 
-
-Parameters:
+.. _multicompartment-models:
```
A reference to multicompartment model documentation should also be linked from `iaf_cond_alpha_mc` neuron model. Or a general documentation needs to be created for multicompartment models which could be then linked from neuron models.

sarakonradi replied:
Done.

clinssen replied:
I would suggest to link this from the "Guides" page on RtD as well.

clinssen commented on models/pp_cond_exp_mc_urbanczik.h:
```diff
@@ -107,82 +107,137 @@ class pp_cond_exp_mc_urbanczik_parameters
+delayed version of the error signal with the presynaptic spike train
+(:math:`S_{pre}`), see panel c).
+
+.. figure:: ../static/img/multicompartment.png
```
Can't find this figure, is it already on the git repository or missing from this PR?

jstapmanns replied:
Hi, the figure should be included in this PR, see /doc/userdoc/static/img/multicompartment.png.

sarakonradi commented:
@pnbabu Thanks for the review. I've applied all your suggestions.

pnbabu reviewed (approved):
Thanks! Looks good to me.

jessica-mitchell commented on models/pp_cond_exp_mc_urbanczik.h:
```diff
@@ -107,82 +107,140 @@ class pp_cond_exp_mc_urbanczik_parameters
+<fig-multicompartment>`. Conversely, if the synaptic weight depends on the
+state of the postsynaptic neuron, the delay also represents the time the
+information on the state propagates back through the dendrite to the synapse.
+
```
"For multicompartment models, this amounts to positioning the delay directly behind the incoming synapse. . ."
@jstapmanns Is the 'positioning of the delay' something the user does when setting up a simulation or is it inherent in the model code itself? As in is this a variable that can be modified by a user in a Python script? If so, I was wondering if along with the figure, it would be a good idea to show a code snippet of a script. So you can see how the biological side is represented in NEST.

jstapmanns replied:
No, the position of the delay is not configurable by the user. It is a basic property of NEST that it is always located on the postsynaptic side.

jessica-mitchell replied:
@jstapmanns Ok, thanks for the clarification; I think it might be helpful reword it (in my reading of it it's not clear who is doing  the 'positioning')
Does this make sense "For multicompartment models in NEST, this means the delay is positioned directly behind the incoming synapse . . ."

jstapmanns replied:
@jessica-mitchell I agree with your suggestion.

jessica-mitchell commented on models/pp_cond_exp_mc_urbanczik.h:
```diff
@@ -107,82 +107,140 @@ class pp_cond_exp_mc_urbanczik_parameters
+might be associated with information traveling through the explicitly modeled
+dendritic compartments.
+
+In case of the Urbanczik synapse, the change of the synaptic weight is driven by
```
```suggestion
In the :ref:`Urbanczik synapse <urbanczik_synapse>`, the change of the synaptic weight is driven by
```

jessica-mitchell commented on models/pp_cond_exp_mc_urbanczik.h:
```diff
@@ -107,82 +107,140 @@ class pp_cond_exp_mc_urbanczik_parameters
+dendritic compartments.
+
+In case of the Urbanczik synapse, the change of the synaptic weight is driven by
+an error signal which is the difference between the firing rate of the soma
```
```suggestion
an error signal, which is the difference between the firing rate of the soma
```
What do you mean by error signal?

jstapmanns replied:
@jessica-mitchell I tried to include the definition of "error signal" in the next half sentence "... error signal which is the difference between the firing rate ..." Apparently, this was not clearly expressed. Would it help to write "... error signal which is defined as the difference between the firing rate ..."? One could also replace the term "error signal" by "prediction error" which is used in [1]. However, as far as I know "error signal" is also quite common in the context of synaptic plasticity.

jessica-mitchell replied:
If error signal is common terminology, then I'm good with it. It was just something I wasn't familiar with.

jstapmanns replied:
Okay, then let's keep "error signal".

jessica-mitchell commented on models/pp_cond_exp_mc_urbanczik.h:
```diff
@@ -107,82 +107,140 @@ class pp_cond_exp_mc_urbanczik_parameters
+In case of the Urbanczik synapse, the change of the synaptic weight is driven by
+an error signal which is the difference between the firing rate of the soma
+(derived from the somatic spike train :math:`S_{post}`) and the
+dendritic prediction of it (derived from the dendritic membrane potential
```
```suggestion
dendritic prediction of the firing rate of the soma (derived from the dendritic membrane potential
```

jstapmanns replied:
@jessica-mitchell  I agree.

jessica-mitchell commented on models/pp_cond_exp_mc_urbanczik.h:
```diff
@@ -107,82 +107,140 @@ class pp_cond_exp_mc_urbanczik_parameters
+:math:`V`). The original publication [1]_ does not assume any delay in the
+interaction between the soma and the dendritic compartment. Therefore, we
+compute the error signal from the firing rate and the dendritic prediction at
+equal time points. Due to the synaptic delay :math:`d`, the synapse combines a
```
What is meant by equal time points?

jstapmanns replied:
@jessica-mitchell  I meant the following: Let's assume that the firing rate is given by `f` and the dendritic prediction is denoted by `p`. Then the error signal `e` is computed as `e(t) = f(t) - p(t)`, where the time arguments of `f` and `p` are equal.

jessica-mitchell replied:
@jstapmanns If this is accepted terminology then I think it can be left as is, but if this might not be clear to a user then this explanation may be warranted.

jstapmanns replied:
@jessica-mitchell Do you think the following wording explains it better: "Therefore, we evaluate the firing rate and the dendritic prediction at equal time points to calculate the error signal at that time point."?

jessica-mitchell replied:
I think so

jessica-mitchell commented on models/pp_cond_exp_mc_urbanczik.h:
```diff
@@ -107,82 +107,140 @@ class pp_cond_exp_mc_urbanczik_parameters
+   :width: 75 %
+
+   a) Two point neurons (red circles *pre* and *post*) connected via a synapse.
+   In NEST, the delay is entirely on the postsynaptic side and in case of point
```
```suggestion
   In NEST, the delay is entirely on the postsynaptic side, and in the case of point
```

jessica-mitchell commented on models/pp_cond_exp_mc_urbanczik.h:
```diff
@@ -107,82 +107,140 @@ class pp_cond_exp_mc_urbanczik_parameters
+
+   a) Two point neurons (red circles *pre* and *post*) connected via a synapse.
+   In NEST, the delay is entirely on the postsynaptic side and in case of point
+   neurons it is interpeted as the dendritic delay. b) Two two-compartment
```
```suggestion
   neurons, it is interpreted as the dendritic delay. b) Two two-compartment
```

jessica-mitchell commented on models/pp_cond_exp_mc_urbanczik.h:
```diff
@@ -107,82 +107,140 @@ class pp_cond_exp_mc_urbanczik_parameters
+   neuron models composed of a somatic (green) and a dendritic (blue)
+   compartment. The soma of the presynaptic neuron is connected to the dendrite
+   of the postsynaptic neuron. The synaptic delay is located behind the synapse
+   and before the dendrite. c) Time trace of the State variables that enter the
```
I am a bit confused by what c) is trying to say. 
Can you be specific as to which variables you are referring to? 
What are postsynaptic quantities - quantity is very vague term.
Is the sentence implying that the dendritic membrane potential and the somatic spike train are the aforementioned postsynaptic quantities? (If so, there should be a colon : rather than semicolon.)
How is the presynaptic spike train _combined with_ a delayed version of postsynaptic quantities? Combined how?

jstapmanns replied:
@jessica-mitchell The variables I am referring to are the presynaptic spike train `S_pre`, the postsynaptic spike train `S_post`, and the dendritic membrane potential `V`. Therefore, yes, the dendritic membrane potential and the somatic spike train are the aforementioned postsynaptic quantities and the semicolon should be replaced by a colon.
Regarding the question how these quantities are combined, I do not think that it should be explained in more detail here since it is done properly in reference [1] which provides all formulas and derivations.

jessica-mitchell replied:
ok, sounds good

jessica-mitchell commented on models/pp_cond_exp_mc_urbanczik.h:
```diff
@@ -107,82 +107,140 @@ class pp_cond_exp_mc_urbanczik_parameters
 Remarks:
 
-The neuron model uses standard units instead of the unitless quantities used in [1].
+The neuron model uses standard units instead of the unitless quantities used in
```
What are standard units?

jstapmanns replied:
@jessica-mitchell With the term "standard units" we refer to the units that are used in NEST in all models, e.g. mV, nS, ms, pA, ...

jessica-mitchell replied:
```suggestion
The neuron model uses standard units of NEST instead of the unitless quantities used in
```

jstapmanns replied:
@jessica-mitchell Yes, I agree.

pnbabu replied:
> I'd suggest to explicitly state that

This suggestion still needs to be addressed.

jessica-mitchell commented on models/pp_cond_exp_mc_urbanczik.h:
```diff
@@ -107,82 +107,140 @@ class pp_cond_exp_mc_urbanczik_parameters
-are stored as C arrays, with index 0 being soma.
+.. note::
+   All parameters that occur for both compartments are stored as C arrays, with
+   index 0 being soma.
```
```suggestion
   index 0 being the soma.
```

jessica-mitchell replied:
Considering this is part of user documentation, does the reader need to know the compartment are stored as C arrays?

jstapmanns replied:
@jessica-mitchell No, this information is only useful for developers, I think.

jessica-mitchell replied:
@jstapmanns  in that case I'd suggest to move that note out of this documentation block

jstapmanns replied:
@jessica-mitchell Yes.

jessica-mitchell commented on models/urbanczik_synapse.h:
```diff
@@ -40,51 +40,66 @@
-that are capable of doing this archiving. So far, the only compatible model is pp_cond_exp_mc_urbanczik.
+urbanczik_synapse is a connector to create Urbanczik synapses as defined in
+[1]_ that can connect suitable :ref:`multicompartment models
+<multicompartment-models>`. In contrast to usual STDP, the change of the
```
```suggestion
<multicompartment-models>`. In contrast to most STDP models, the change of the
```

jessica-mitchell commented on models/urbanczik_synapse.h:
```diff
@@ -40,51 +40,66 @@
+urbanczik_synapse is a connector to create Urbanczik synapses as defined in
+[1]_ that can connect suitable :ref:`multicompartment models
+<multicompartment-models>`. In contrast to usual STDP, the change of the
+synaptic weight does not only depend on the pre- and postsynaptic spike timing
```
```suggestion
synaptic weight depends on the postsynaptic dendritic potential
```

jstapmanns replied:
@jessica-mitchell What we wanted to express here is the following: Ordinary STDP depends on the pre- and postsynaptic spike timing. Urbanczik-Senn plasticity additionally depends on the dendritic membrane potential of the postsynaptic neuron.

jessica-mitchell replied:
hmm I seem to have messed up my edit, but I understand what you mean - I was trying to remove the 'not' in the phrase and say " . . . synaptic weight depends on the postsynaptic dendritic potential, in addition to the pre- and postsynaptic spike timing". But upon thinking of this further I think the way you had it might be better in this case.

jstapmanns replied:
@jessica-mitchell I see, but I like your suggestion, so let's replace the sentence with what you wrote.

jessica-mitchell commented on models/urbanczik_synapse.h:
```diff
@@ -40,51 +40,66 @@
+but also on the postsynaptic dendritic potential.
 
-Parameters:
+Urbanczik synapses require archiving of continuous quantities. Therefore they
```
Is there a better word than quantity?

jstapmanns replied:
@jessica-mitchell Yes. To be more precise one could write: Urbanczik synapses require the archiving of the dendritic membrane potential which is continuous in time. Therefore ..."

jessica-mitchell replied:
that sounds better

jstapmanns replied:
@jessica-mitchell Okay, then I will add that.

pnbabu replied:
Could this be reworded as suggested?

jessica-mitchell commented on models/urbanczik_synapse.h:
```diff
@@ -40,51 +40,66 @@
+++++++++++
 
-References:
+.. [1] R. Urbanczik, W. Senn (2014). Learning by the Dendritic Prediction of
```
```suggestion
.. [1] Urbanczik R. and Senn W (2014). Learning by the dendritic prediction of
```

jessica-mitchell commented on models/urbanczik_synapse.h:
```diff
@@ -40,51 +40,66 @@
 
-References:
+.. [1] R. Urbanczik, W. Senn (2014). Learning by the Dendritic Prediction of
+       Somatic Spiking. Neuron, 81, 521 - 528.
```
```suggestion
       somatic spiking. Neuron, 81:521 - 528. https://doi.org/10.1016/j.neuron.2013.11.030
```

jessica-mitchell reviewed (changes requested):
@sarakonradi @jstapmanns  thanks for your work on this! I have a few changes and also some questions, where I was not clear to the meaning of text

jstapmanns commented on models/pp_cond_exp_mc_urbanczik.h:
```diff
@@ -107,82 +107,140 @@ class pp_cond_exp_mc_urbanczik_parameters
-Each compartment can also receive current input from a current generator, and an external
-(rheobase) current can be set for each compartment.
+The model has two compartments: soma and dendrite, labeled as s and p,
+respectively. There is one excitatory and one inhibitory conductance-based
```
@jessica-mitchell This sentence is somewhat misleading and we should remove it. Instead, I suggest adding two sentences a few lines further down.

jstapmanns commented on models/pp_cond_exp_mc_urbanczik.h:
```diff
@@ -107,82 +107,140 @@ class pp_cond_exp_mc_urbanczik_parameters
 Synapses, including those for injection external currents, are addressed through
 the receptor types given in the receptor_types entry of the state dictionary.
 Note that in contrast to the single-compartment models, all
 synaptic weights must be positive numbers!
```
Here one could clarify this a bit further by adding: "The distinction between excitatory and inhibitory synapses is made explicitly by specifying the receptor type of the synapse. For example, `'receptor_type'='dendritic_exc'` results in an excitatory input and `'receptor_type'='dendritic_inh'` results in an inhibitory input to the dendritic compartment."

jessica-mitchell replied:
@jstapmanns  Thanks for the added clarification. Could you add this to the text?

jstapmanns replied:
@jessica-mitchell Yes, I will take care of that.

jessica-mitchell commented on models/pp_cond_exp_mc_urbanczik.h:
```diff
@@ -105,82 +105,140 @@ class pp_cond_exp_mc_urbanczik_parameters
-(rheobase) current can be set for each compartment.
+The model has two compartments: soma and dendrite, labeled as s and p,
+respectively. There is one excitatory and one inhibitory conductance-based
+conductance-based synapse. Each compartment can also receive current input
```
```suggestion
Each compartment can also receive current input
```

sarakonradi commented:
Thanks for the reviews. I have merged @jstapmanns's PR ([#11](https://github.com/sarakonradi/nest-simulator/pull/11)), which addresses @jessica-mitchell's suggestions.

jessica-mitchell commented on models/pp_cond_exp_mc_urbanczik.h:
```diff
@@ -143,26 +146,28 @@ assumed to be located entirely at the postsynaptic side. For point neurons, it
+synapse.
+
+For multicompartment models in NEST, this means the delay is positioned directly
+behind the incoming synapse, i.e., before the first dendritic compartment on the
```
```suggestion
behind the incoming synapse, that is, before the first dendritic compartment on the
```

jessica-mitchell commented on models/pp_cond_exp_mc_urbanczik.h:
```diff
@@ -143,26 +146,28 @@ assumed to be located entirely at the postsynaptic side. For point neurons, it
-delayed version of the error signal with the presynaptic spike train
-(:math:`S_{pre}`), see :ref:`panel c) <fig-multicompartment>`.
+In the :ref:`Urbanczik synapse <urbanczik_synapse>`, the change of the synaptic
+weight is driven by an error signal which is the difference between the firing
```
```suggestion
weight is driven by an error signal, which is the difference between the firing
```

jessica-mitchell reviewed (approved):
Thanks @jstapmanns  and @sarakonradi. I think this looks good. Just had a couple of minor changes but I approve the PR

jstapmanns reviewed (approved):
Thank you @sarakonradi and @jessica-mitchell preparing and improving this PR.

sarakonradi commented:
@pnbabu I just merged another PR ([#12](https://github.com/sarakonradi/nest-simulator/pull/12)) by @jstapmanns. Please check if we have addressed your concerns. Thanks!

pnbabu reviewed (approved):
Thanks. Looks good to me.
